---
title: "A brief introduction to countseq package"
author:
-   name: "Davide Raffaelli"
    affiliation: Politecnico di Milano
    email: davide.raffaellii@gmail.com
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{A brief introduction to countseq package}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'asis'}
library(BiocStyle)
```
```{r, echo = FALSE}
library(knitr)
```

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

```{r setup}
library(countseq)
```

# Introduction
The `r Rpackage("countseq")` package provides functions that allows you to get
overall count of sequences of DNA that match in regions of a genome and other
useful functions to manage and store this data.

# Functions

## Get matrix of overall counts

Get matrix of **overall counts** is simple and require only one function: 
**countSeqsMatrix**. You have to pass a genome, the regions and the sequences.
This example want to check if there could be a CpG island at the promoters of
chromosome 22 genes.

```{r, message=FALSE}
 #Check if there are CpG island in gene's promoters of chromosome 22 in human
 #Charge Hsapiens genome
 library(BSgenome.Hsapiens.UCSC.hg38)

 #Charge file where are stored genomic positions of TSS in chr22
 df <- read.table(system.file("extdata", "chr22_unique_txStart.txt",
     package = "countseq"), header = TRUE)
 
 #Get a GRanges object which contains regions of CpG island
 library(GenomicRanges)
 regs <-GRanges("chr22", 
     IRanges(start = df$txStart -130, end = df$txStart -30))

 #Get a DNAStringSet which contains the sequences to be mapped
 library(Biostrings)
 seqs <- DNAStringSet(c("CG", "AT", "TA"))

 #Calculate the matrix
 mat <- countSeqsMatrix(BSgenome.Hsapiens.UCSC.hg38::Hsapiens, regs, seqs)
 
 #Percentage of promoters that are likely to contain a CpG island
 sum(mat[,"CG"] > mat[,"AT"] & mat[,"CG"] > mat[,"TA"])/nrow(mat)
```

## Sort the matrix obtained from countSeqsMatrix

If you need to **sort** the matrix of overall counts, because you have to do 
some operations which require a ordered matrix, this package provide the 
function **sortMatrix**. You have to just pass the matrix and tell you want the 
matrix sorted in decreasing or increasing way.

```{r}
#Create matrix
matSeqs <- matrix(data = c(0,3,1,0,20,13), nrow = 2)
rownames(matSeqs) <- c("chr1:1000001", "chr1:1000101")
colnames(matSeqs) <- c("seq1", "seq2", "seq3")

#Sort matrix in descreasign way
sortMatrix(matSeqs, TRUE)
```

## Delete rows and columns full of zeros

This package also provide the **reduceMatrix** method to delete rows and columns
full of zeros, in order to exclude them from your calculations or to save space.

```{r}
#Create matrix
matSeqs <- matrix(data = c(0,3,1,0,20,13), nrow = 2)
rownames(matSeqs) <- c("chr1:1000001", "chr1:1000101")
colnames(matSeqs) <- c("seq1", "seq2", "seq3")

#Reduce matrix
reduceMatrix(matSeqs)
```

## Save and get matrix from a file

In order to save the matrix that you obtained and to retrive it, this package
allows you to save the matrix of overall counts in a **.txt file** with 
**writeToFile** function. You can also read this file in order to get the 
matrix when you need, with **readFromFile** function.

```{r}
#Create matrix
matSeqs <- matrix(data = c(0,3,1,0,20,13), nrow = 2)
rownames(matSeqs) <- c("chr1:1000001", "chr1:1000101")
colnames(matSeqs) <- c("seq1", "seq2", "seq3")

#Save matrix in a text file
writeToFile(matSeqs, "matrix.txt")

#Get matrix from file
m <- readFromFile("matrix.txt", integers = TRUE)
m

#remove file
file.remove("matrix.txt")

```
